query-daylight-dplyr.Rmd

```diff
@@ -0,0 +1,327 @@
+---
+title: "SQL Window Functions"
+author:
+  - name: Jonathan Keane
+    email: \email{jkeane@gmail.com}
+output:
+  html_document
+---
+
+```{r, echo=FALSE, message=FALSE, results='hide'}
+library(dplyr)
+daylight <- read.table("daylight.tsv", sep = "\t", col.names = c("month", "day", "light"))
+```
+
+See the data
+```{r}
+print(daylight)
+```
+
+how many hours of daylight on a particular day (for example December 1)?
+```{r}
+daylight %>%
+  filter(month == 12, day == 1)
+```
+
+A visualization of this data:
+![2017 Daily Sunrise and Sunset at Raleigh-Durham International Airport, North Carolina, United States](https://cloud.githubusercontent.com/assets/280509/23589002/dab71d0a-0193-11e7-8e91-33688ba060fd.png)
+
+standard aggregation example: max daylight for each month 
+```{r}
+daylight %>%
+  group_by(month) %>% 
+  summarise(max(light)) %>% 
+  arrange(month)
+```
+groups sets of rows together into single rows
+individual row values are not included in the output
+
+which day in each month does the max occur on?
+
+simple window function example: max daylight for each month
+```{r}
+daylight %>% 
+  group_by(month) %>% 
+  mutate(month_max = max(light)) %>% 
+  arrange(month, day)
+```
+window functions (aka analytic functions) are applied over sets of rows without combining them
+individual row values can be included in the output
+
+return only the row for the day with the max daylight in each month
+
+doesn't work; window expressions not allowed in WHERE clause
+```{sql, connection=con, eval=FALSE}
+SELECT month, day, light
+  FROM daylight
+ 	WHERE light = MAX(light) OVER(PARTITION BY month);
+```
+
+workaround: put it in the SELECT list
```

Comment:
Can we delete this part? I think it's relevant to SQL but not to dplyr
```suggestion
```

Reply:
I like the parallelism, though maybe it would be better to make that explicit?

Reply:
For now I'm going to merge this into a new branch of this repo named `r`. Later maybe we can come back and tidy up everything and merge together the branches.